rasterio/__init__.py

```diff
@@ -333,7 +318,18 @@ def func(*args, **kwds):
     # which can be converted to a string.
     else:
         raw_dataset_path = os.fspath(fp)
-        path = _parse_path(raw_dataset_path)
+        stack = ExitStack()
```

Comment:
Should stack be defined above the `if` statement?

Reply:
The following code is simpler if it's defined here and attached to every dataset coming out of the next if/else (line 334+). If there's no opener, it's an empty stack, with very small cost of closing.

Reply:
I see why I was confused. The `elif` statements above should be just `if` because they always `return` or raise an exception. If that is changed, there will be no need for the `else` block here and `stack` is safe to be defined here.